{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "initial_id",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import all libraries\n",
    "import pandas as pd\n",
    "import requests\n",
    "import bs4\n",
    "import re\n",
    "import logging\n",
    "import concurrent.futures\n",
    "import threading\n",
    "from time import sleep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "6c2e65b93bcafb3c",
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def get_links():\n",
    "    urls = list(pd.read_csv('books_url.csv')['link'])\n",
    "    return urls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "fe12b9fe14593376",
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def get_response(input_url):\n",
    "    headers = {\n",
    "        'User-Agent': 'My User Agent 1.0',\n",
    "        \"Accept-Language\": \"en-US,en;q=0.5\"\n",
    "    }\n",
    "    response = requests.get(input_url, headers=headers)\n",
    "    return response"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "dd775c4290d33dfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_soup(input_url):\n",
    "    headers = {\n",
    "        'User-Agent': 'My User Agent 1.0',\n",
    "        \"Accept-Language\": \"en-US,en;q=0.5\"\n",
    "    }\n",
    "    response = requests.get(input_url, headers=headers)\n",
    "    if response.status_code != 200:\n",
    "        print(\"Error in getting link\")\n",
    "        print(\"response code is : \", response.status_code)\n",
    "    soup = bs4.BeautifulSoup(response.text, 'html.parser')\n",
    "    return soup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "efc3968719f0d788",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_fa_title(soup):\n",
    "    title = soup.select('.product-name strong')[0]\n",
    "    return title.text\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "2c4dab072bbbc2ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_en_title(soup):\n",
    "    title = soup.select('.product-name-englishname')[0]\n",
    "    return title.text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "5c6151c0ff09fc3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_price(soup):\n",
    "    price = soup.select('.price-broken , .col-md-7 .price:nth-child(1)')[0].text\n",
    "    return int(price.replace(',', ''))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "20dfbbc77ef0ea67",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_discount(soup):\n",
    "    try:\n",
    "        discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
    "        discount_price = ((get_price(soup) - discount_price) / get_price(soup)) * 100\n",
    "    except Exception:\n",
    "        discount_price = 0\n",
    "        logging.exception(\"This book has no discount!\")\n",
    "    return discount_price"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "3b95900da3f89757",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_score(soup):\n",
    "    soup = soup.find('div', {'class': 'col-md-7'}).find('li', {'class': 'pull-left'}).find('div',\n",
    "                                                                                           {'class': 'my-rating'})\n",
    "    soup_str = str(soup)\n",
    "\n",
    "    match = re.search(r'data-rating=\"(\\d+\\.\\d+)\"', soup_str)\n",
    "    if match:\n",
    "        data_rating = match.group(1)\n",
    "        return data_rating\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "3c5c9ddc5b07b81e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_publisher(soup):\n",
    "    try:\n",
    "        publisher_tag = soup.select('div.prodoct-attribute-items:nth-child(1) > a')[0]\n",
    "        publisher_link=publisher_tag.get('href')\n",
    "        publisher_id=publisher_link.split('/')[2].split('-')[0]\n",
    "        publisher_name=publisher_tag.text.strip()\n",
    "    except Exception:\n",
    "        publisher_link=-1\n",
    "        publisher_id=-1\n",
    "        publisher_name=-1\n",
    "    return {'id':publisher_id,'name':publisher_name,'link':publisher_link}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "235a6bc598b6b7ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_author(soup):\n",
    "    authors_list=[]          \n",
    "    try:\n",
    "        authors_a_tag = soup.select('.prodoct-attribute-items+ .prodoct-attribute-items > a')\n",
    "        if(len(authors_a_tag))==0:\n",
    "                return authors_list\n",
    "        for author_a_tag in authors_a_tag:\n",
    "                author_link=author_a_tag.get('href')\n",
    "                author_id=author_link.split('/')[2].split('-')[0]\n",
    "                author_name=author_a_tag.text.strip()\n",
    "                authors_list.append({'id':author_id,'name':author_name,'link':author_link})\n",
    "    except Exception:\n",
    "        author = -1\n",
    "    return authors_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "b21b1fa78bdd0d0e",
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def is_author_available(soup):\n",
    "    try:\n",
    "        existence = soup.select('.pull-left+ li span')[0].text\n",
    "    except:\n",
    "        existence = None\n",
    "        logging.exception(\"This book has no author!\")\n",
    "    return existence\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "91a45fd031eab930",
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def get_book_attribute(soup):\n",
    "    rows = soup.find('table', {'class': 'product-table'}).findAll('td')\n",
    "    code = -1\n",
    "    isbn = -1\n",
    "    size = -1\n",
    "    pages = -1\n",
    "    per_cal = -1\n",
    "    ad_cal = -1\n",
    "    material = -1\n",
    "    series = -1\n",
    "    send_time = -1\n",
    "    language = 'فارسی'\n",
    "    code_flag = 0\n",
    "    isbn_flag = 0\n",
    "    size_flag = 0\n",
    "    pages_flag = 0\n",
    "    per_cal_flag = 0\n",
    "    ad_cal_flag = 0\n",
    "    material_flag = 0\n",
    "    language_flag = 0\n",
    "    series_flag = 0\n",
    "    send_time_flag = 0\n",
    "\n",
    "    for row in rows:\n",
    "        text = row.text.strip()\n",
    "        if code_flag == 1:\n",
    "            code = int(text)\n",
    "            code_flag = 0\n",
    "        elif isbn_flag == 1:\n",
    "            isbn = text\n",
    "            isbn = re.sub('[^0-9-]', '', isbn)\n",
    "            isbn_flag = 0\n",
    "        elif size_flag == 1:\n",
    "            size = text\n",
    "            size_flag = 0\n",
    "        elif pages_flag == 1:\n",
    "            pages = int(text)\n",
    "            pages_flag = 0\n",
    "        elif per_cal_flag == 1:\n",
    "            per_cal = int(text)\n",
    "            per_cal_flag = 0\n",
    "        elif ad_cal_flag == 1:\n",
    "            ad_cal = int(text)\n",
    "            ad_cal_flag = 0\n",
    "        elif material_flag == 1:\n",
    "            material = text\n",
    "            material_flag = 0\n",
    "        elif language_flag == 1:\n",
    "            language = text\n",
    "            language_flag = 0\n",
    "        elif series_flag == 1:\n",
    "            series = int(text)\n",
    "            series_flag = 0\n",
    "        elif send_time_flag == 1:\n",
    "            send_time = text\n",
    "            send_time_flag = 0\n",
    "\n",
    "        if 'کد کتاب' in text:\n",
    "            code_flag = 1\n",
    "        elif 'شابک' in text:\n",
    "            isbn_flag = 1\n",
    "        elif 'قطع' in text:\n",
    "            size_flag = 1\n",
    "        elif 'تعداد صفحه' in text:\n",
    "            pages_flag = 1\n",
    "        elif 'سال انتشار شمسی' in text:\n",
    "            per_cal_flag = 1\n",
    "        elif 'سال انتشار میلادی' in text:\n",
    "            ad_cal_flag = 1\n",
    "        elif 'نوع جلد' in text:\n",
    "            material_flag = 1\n",
    "        elif 'زبان کتاب' in text:\n",
    "            language_flag = 1\n",
    "        elif 'سری چاپ' in text:\n",
    "            series_flag = 1\n",
    "        elif 'زودترین زمان ارسال' in text:\n",
    "            send_time_flag = 1\n",
    "\n",
    "    return [code, isbn, size, pages, per_cal, ad_cal, material, language, series, send_time]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "5c297eef97f3ef6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_summary(soup):\n",
    "    summary = soup.select('.product-description')[0].text.strip()\n",
    "    return summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "9680a41c872e4ad7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_tags(soup):\n",
    "    tags = soup.select('.product-tags-item')\n",
    "    tags_list = []\n",
    "    for tag in tags:\n",
    "        tags_list += [tag.text.strip()]\n",
    "    return tags_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "b487ed5aba32bcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_book_detail(book_soup, site_index):\n",
    "    book_fa_title = get_fa_title(book_soup)\n",
    "    book_en_title = get_en_title(book_soup)\n",
    "    book_price = get_price(book_soup)\n",
    "    book_discount_percent = get_discount(book_soup)\n",
    "    book_score = get_score(book_soup)\n",
    "    book_publisher = get_publisher(book_soup)\n",
    "    book_author = get_author(book_soup)\n",
    "    book_author_presence = is_author_available(book_soup)\n",
    "\n",
    "    [book_code, book_isbn, book_size, book_pages, book_publication_per_date, book_publication_ad_date,\n",
    "     book_cover_material, book_language, book_print_series, book_earliest_send_time] = get_book_attribute(book_soup)\n",
    "\n",
    "    book_data = [site_index, int(book_code), book_isbn, book_fa_title, book_en_title, book_price,\n",
    "                 int(book_discount_percent), book_score, book_publisher, book_author, int(book_pages),\n",
    "                 int(book_publication_per_date), int(book_publication_ad_date), book_size, book_cover_material,\n",
    "                 book_language, int(book_print_series), book_earliest_send_time, book_author_presence]\n",
    "    return book_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "52bb9e3fa91b9548",
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def get_book_site_summary(book_soup, site_index):\n",
    "    try:\n",
    "        book_summary = get_summary(book_soup)\n",
    "    except Exception:\n",
    "        book_summary = None\n",
    "        logging.exception(\"This book has no summary!\")\n",
    "    return [site_index, book_summary]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "12597cd1c3c3e9f5",
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def get_book_site_tags(book_soup, site_index):\n",
    "    book_tags = get_tags(book_soup)\n",
    "    book_tags_list = []\n",
    "    for tag in book_tags:\n",
    "        book_tags_list += [[site_index, tag]]\n",
    "    return book_tags_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "33fc61291443bf82",
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def get_site_awards(soup, site_index):\n",
    "    awards_list = []\n",
    "    awards = soup.select('book_soup, site_index')\n",
    "\n",
    "    for award in awards:\n",
    "        print(award.text)\n",
    "        awards += [award.text]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "3e410df15deddef2",
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def get_req_list(list, req_count):\n",
    "    if len(list) >= req_count:\n",
    "        request_list = list[:req_count].copy()\n",
    "    else:\n",
    "        request_list = list.copy()\n",
    "    return request_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "6765fb21ffc5eee4",
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def scrape(site_soup):\n",
    "    try:\n",
    "        with lock:\n",
    "            global site_index\n",
    "            site_summary_data_list.append(get_book_site_summary(site_soup, site_index))\n",
    "            site_tags_data_list.extend(get_book_site_tags(site_soup, site_index))\n",
    "            site_page_books = site_soup.select('.clearfix .clearfix .row')\n",
    "            for book_index in range(0, len(site_page_books), 2):\n",
    "                data = get_book_detail(site_page_books[book_index], site_index)\n",
    "                books_data_list.append(data)\n",
    "            site_index += 1\n",
    "    except Exception:\n",
    "        logging.exception(\"An error occurred\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "113b8344e2efc33f",
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def fast_scrape(link):\n",
    "    try:\n",
    "        site_soup = get_soup(link)\n",
    "        with lock:\n",
    "            global site_index\n",
    "            site_summary_data_list.append(get_book_site_summary(site_soup, site_index))\n",
    "            site_tags_data_list.extend(get_book_site_tags(site_soup, site_index))\n",
    "            site_page_books = site_soup.select('.clearfix .clearfix .row')\n",
    "            for book_index in range(0, len(site_page_books), 2):\n",
    "                data = get_book_detail(site_page_books[book_index], site_index)\n",
    "                writers_data_list.extend(data[9])     #9th index is the writer column which is a list of writers\n",
    "                publishers_data_list.append(data[8])  #8th column is dict of publisher\n",
    "                data[8]=data[8]['id']                 #convert 8th column from dict to the publisher's id\n",
    "                writers_list_of_dict=data[9]          #writes list which is a list of dictionary\n",
    "                data.pop(9)                           #remove 9th column from data(9th column was writers)\n",
    "                for w_id in writers_list_of_dict:\n",
    "                     books_writers_data_list.append({'book_id':data[1],'writer_id':w_id['id']})\n",
    "                books_data_list.append(data)\n",
    "            site_index += 1\n",
    "    except Exception:\n",
    "        logging.exception(\"An error occurred\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba91c7d84afba5a4",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "<h1>Detailed Scraper</h1>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "9a609902c961aa8e",
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001B[1;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[1;31mKeyboardInterrupt\u001B[0m                         Traceback (most recent call last)",
      "\u001B[1;32mc:\\Users\\Asus\\Desktop\\Booktopia\\scrap.ipynb Cell 24\u001B[0m in \u001B[0;36m<module>\u001B[1;34m\u001B[0m\n\u001B[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/Asus/Desktop/Booktopia/scrap.ipynb#X32sZmlsZQ%3D%3D?line=13'>14</a>\u001B[0m book_urls \u001B[39m=\u001B[39m links\u001B[39m.\u001B[39mcopy()\n\u001B[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/Asus/Desktop/Booktopia/scrap.ipynb#X32sZmlsZQ%3D%3D?line=15'>16</a>\u001B[0m \u001B[39mwhile\u001B[39;00m \u001B[39mlen\u001B[39m(book_urls):\n\u001B[1;32m---> <a href='vscode-notebook-cell:/c%3A/Users/Asus/Desktop/Booktopia/scrap.ipynb#X32sZmlsZQ%3D%3D?line=16'>17</a>\u001B[0m     sleep(sleep_time)  \u001B[39m#sleep so that the site does not ban us\u001B[39;00m\n\u001B[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/Asus/Desktop/Booktopia/scrap.ipynb#X32sZmlsZQ%3D%3D?line=17'>18</a>\u001B[0m     request_list \u001B[39m=\u001B[39m get_req_list(book_urls, book_count_request)  \u001B[39m#list of book's urls we want to send request \u001B[39;00m\n\u001B[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/Asus/Desktop/Booktopia/scrap.ipynb#X32sZmlsZQ%3D%3D?line=18'>19</a>\u001B[0m     \u001B[39mwith\u001B[39;00m concurrent\u001B[39m.\u001B[39mfutures\u001B[39m.\u001B[39mThreadPoolExecutor(max_workers\u001B[39m=\u001B[39mmax_threads) \u001B[39mas\u001B[39;00m executor:\n",
      "\u001B[1;31mKeyboardInterrupt\u001B[0m: "
     ]
    }
   ],
   "source": [
    "links = get_links()[:200] + ['https://www.iranketab.ir/book/270-gone-with-the-wind']\n",
    "\n",
    "page_response = []\n",
    "books_data_list = []\n",
    "site_tags_data_list = []\n",
    "site_summary_data_list = []\n",
    "\n",
    "site_index = 1\n",
    "sleep_time = 0.5\n",
    "max_threads = 20\n",
    "book_count_request = 20  #number of requests per time\n",
    "\n",
    "lock = threading.Lock()\n",
    "book_urls = links.copy()\n",
    "\n",
    "while len(book_urls):\n",
    "    sleep(sleep_time)  #sleep so that the site does not ban us\n",
    "    request_list = get_req_list(book_urls, book_count_request)  #list of book's urls we want to send request \n",
    "    with concurrent.futures.ThreadPoolExecutor(max_workers=max_threads) as executor:\n",
    "        future_list = executor.map(get_response, request_list)\n",
    "        for future in future_list:\n",
    "            try:\n",
    "                data = future\n",
    "                page_response.append(data)\n",
    "            except Exception as exc:\n",
    "                continue\n",
    "        for item in page_response:\n",
    "            if item.status_code == 200:\n",
    "                page_url = item.url\n",
    "                if page_url in request_list:\n",
    "                    page_soup = bs4.BeautifulSoup(item.content, 'html.parser')\n",
    "                    scrape(page_soup)\n",
    "                    book_urls.remove(page_url)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6191d96283542fa",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "<h1>Fast Scraper</h1>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "8d0c478f6decda4a",
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 10, in fast_scrape\n",
      "    data = get_book_detail(site_page_books[book_index], site_index)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\2350678637.py\", line 3, in get_book_detail\n",
      "    book_en_title = get_en_title(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1524599449.py\", line 2, in get_en_title\n",
      "    title = soup.select('.product-name-englishname')[0]\n",
      "IndexError: list index out of range\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no discount!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1188541366.py\", line 3, in get_discount\n",
      "    discount_price = int(soup.select('.col-md-12+ .clearfix .price-special')[0].text.replace(',', ''))\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 10, in fast_scrape\n",
      "    data = get_book_detail(site_page_books[book_index], site_index)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\2350678637.py\", line 3, in get_book_detail\n",
      "    book_en_title = get_en_title(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1524599449.py\", line 2, in get_en_title\n",
      "    title = soup.select('.product-name-englishname')[0]\n",
      "IndexError: list index out of range\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:This book has no summary!\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\1156613770.py\", line 3, in get_book_site_summary\n",
      "    book_summary = get_summary(book_soup)\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\3952702039.py\", line 2, in get_summary\n",
      "    summary = soup.select('.product-description')[0].text.strip()\n",
      "IndexError: list index out of range\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n",
      "ERROR:root:An error occurred\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Asus\\AppData\\Local\\Temp\\ipykernel_32944\\581286483.py\", line 14, in fast_scrape\n",
      "    data[9]=data[9]['id']\n",
      "TypeError: list indices must be integers or slices, not str\n"
     ]
    }
   ],
   "source": [
    "links = get_links()[:200] + ['https://www.iranketab.ir/book/270-gone-with-the-wind']\n",
    "\n",
    "books_data_list = []\n",
    "site_tags_data_list = []\n",
    "site_summary_data_list = []\n",
    "writers_data_list = []\n",
    "publishers_data_list = []\n",
    "books_writers_data_list = []\n",
    "\n",
    "site_index = 1\n",
    "max_threads = 20\n",
    "\n",
    "lock = threading.Lock()\n",
    "\n",
    "with concurrent.futures.ThreadPoolExecutor(max_workers=max_threads) as executor:\n",
    "    executor.map(fast_scrape, links)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e8d25f005a8c2c8",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "<h1>Check Completnes</h1>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61f7e5ebf2342413",
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'book_urls' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001B[1;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[1;31mNameError\u001B[0m                                 Traceback (most recent call last)",
      "\u001B[1;32mc:\\Users\\Asus\\Desktop\\Booktopia\\scrap.ipynb Cell 28\u001B[0m in \u001B[0;36m<module>\u001B[1;34m\u001B[0m\n\u001B[1;32m----> <a href='vscode-notebook-cell:/c%3A/Users/Asus/Desktop/Booktopia/scrap.ipynb#X36sZmlsZQ%3D%3D?line=0'>1</a>\u001B[0m \u001B[39mif\u001B[39;00m \u001B[39mlen\u001B[39m(book_urls) \u001B[39m==\u001B[39m \u001B[39m0\u001B[39m:\n\u001B[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/Asus/Desktop/Booktopia/scrap.ipynb#X36sZmlsZQ%3D%3D?line=1'>2</a>\u001B[0m     \u001B[39mprint\u001B[39m(\u001B[39m'\u001B[39m\u001B[39mAll links scraped!\u001B[39m\u001B[39m'\u001B[39m)\n\u001B[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/Asus/Desktop/Booktopia/scrap.ipynb#X36sZmlsZQ%3D%3D?line=2'>3</a>\u001B[0m \u001B[39melse\u001B[39;00m:\n",
      "\u001B[1;31mNameError\u001B[0m: name 'book_urls' is not defined"
     ]
    }
   ],
   "source": [
    "if len(book_urls) == 0:\n",
    "    print('All links scraped!')\n",
    "else:\n",
    "    print('Something wrong happened!',len(book_urls),)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23a1d4d155e402ba",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "<h1>Make Dataframes</h1>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "ad990b979ac7f334",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site_index</th>\n",
       "      <th>code</th>\n",
       "      <th>Isbn</th>\n",
       "      <th>fa_title</th>\n",
       "      <th>en_title</th>\n",
       "      <th>price</th>\n",
       "      <th>discount</th>\n",
       "      <th>score</th>\n",
       "      <th>publisher</th>\n",
       "      <th>author</th>\n",
       "      <th>pages</th>\n",
       "      <th>publication_per_date</th>\n",
       "      <th>publication_ad_date</th>\n",
       "      <th>size</th>\n",
       "      <th>cover_material</th>\n",
       "      <th>print_series</th>\n",
       "      <th>earliest_send_time</th>\n",
       "      <th>presence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [site_index, code, Isbn, fa_title, en_title, price, discount, score, publisher, author, pages, publication_per_date, publication_ad_date, size, cover_material, print_series, earliest_send_time, presence]\n",
       "Index: []"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tableOfData = pd.DataFrame(books_data_list,\n",
    "                           columns=['site_index', 'code', 'Isbn', 'fa_title', 'en_title', 'price', 'discount', 'score',\n",
    "                                    'publisher', 'author', 'pages', 'publication_per_date', 'publication_ad_date',\n",
    "                                    'size', 'cover_material', 'print_series', 'earliest_send_time', 'presence'])\n",
    "tableOfData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15f94ddb4454136d",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_path = \"bookData.csv\"\n",
    "tableOfData.to_csv(file_path, index=False, encoding='utf-8')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49865f3f491e2893",
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site_index</th>\n",
       "      <th>summary</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>کتاب «تکنولوژی های انقلابی» اثری نوشته «آدام گ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>«خواندن کتاب اعتماد کارساز است! نه تنها بسیار ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>طب چینی یا سوزنی یکی از دو رشته درمانی طبی دنی...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>دفتر يادداشت تركيبي پنگوئن:خط دار،بي خط،نقطه ا...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>195</td>\n",
       "      <td>نویسنده مدعی است داستانهایی نوشته است که به نو...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>196</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>197</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>198</td>\n",
       "      <td>کتاب زندگانی من و روزگار سخت، اثری است از جیمز...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>199</td>\n",
       "      <td>کتاب برباد رفته، رمانی است که مارگارت میچل آن ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>201 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     site_index                                            summary\n",
       "0             1  کتاب «تکنولوژی های انقلابی» اثری نوشته «آدام گ...\n",
       "1             2                                               None\n",
       "2             3  «خواندن کتاب اعتماد کارساز است! نه تنها بسیار ...\n",
       "3             4  طب چینی یا سوزنی یکی از دو رشته درمانی طبی دنی...\n",
       "4             5  دفتر يادداشت تركيبي پنگوئن:خط دار،بي خط،نقطه ا...\n",
       "..          ...                                                ...\n",
       "196         195  نویسنده مدعی است داستانهایی نوشته است که به نو...\n",
       "197         196                                               None\n",
       "198         197                                               None\n",
       "199         198  کتاب زندگانی من و روزگار سخت، اثری است از جیمز...\n",
       "200         199  کتاب برباد رفته، رمانی است که مارگارت میچل آن ...\n",
       "\n",
       "[201 rows x 2 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tableOfSummaryData = pd.DataFrame(site_summary_data_list, columns=['site_index', 'summary'])\n",
    "tableOfSummaryData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7181915a808b0a33",
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "file_path = \"BookSummaryData.csv\"\n",
    "tableOfData.to_csv(file_path, index=False, encoding='utf-8')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fb17a40b865591",
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site_index</th>\n",
       "      <th>tag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ادبیات آمریکا</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ادبیات معاصر</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>دهه 2010 میلادی</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>علمی</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>تکنولوژی</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1071</th>\n",
       "      <td>199</td>\n",
       "      <td>فهرست برترین رمان های تاریخی</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1072</th>\n",
       "      <td>199</td>\n",
       "      <td>برترین آثار داستانی با شخصیت اصلی زن</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1073</th>\n",
       "      <td>199</td>\n",
       "      <td>برترین آثار تبدیل شده به فیلم و سریال</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1074</th>\n",
       "      <td>199</td>\n",
       "      <td>فهرست برترین رمان های عاشقانه</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1075</th>\n",
       "      <td>199</td>\n",
       "      <td>فهرست برترین کتاب ها درباره نژادپرستی</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1076 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      site_index                                    tag\n",
       "0              1                          ادبیات آمریکا\n",
       "1              1                           ادبیات معاصر\n",
       "2              1                        دهه 2010 میلادی\n",
       "3              1                                   علمی\n",
       "4              1                               تکنولوژی\n",
       "...          ...                                    ...\n",
       "1071         199           فهرست برترین رمان های تاریخی\n",
       "1072         199   برترین آثار داستانی با شخصیت اصلی زن\n",
       "1073         199  برترین آثار تبدیل شده به فیلم و سریال\n",
       "1074         199          فهرست برترین رمان های عاشقانه\n",
       "1075         199  فهرست برترین کتاب ها درباره نژادپرستی\n",
       "\n",
       "[1076 rows x 2 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tableOfSiteTagsData = pd.DataFrame(site_tags_data_list, columns=['site_index', 'tag'])\n",
    "tableOfSiteTagsData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "807ebe5799696472",
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "file_path = \"bookTagsData.csv\"\n",
    "tableOfData.to_csv(file_path, index=False, encoding='utf-8')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67d3b338aab9ca3b",
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>link</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>72</td>\n",
       "      <td>کتاب آمه</td>\n",
       "      <td>/publisher/72-%da%a9%d8%aa%d8%a7%d8%a8-%d8%a2%...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1588</td>\n",
       "      <td>اشک</td>\n",
       "      <td>/publisher/1588-%d8%a7%d8%b4%da%a9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1217</td>\n",
       "      <td>دنیای اقتصاد</td>\n",
       "      <td>/publisher/1217-%d8%af%d9%86%db%8c%d8%a7%db%8c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1589</td>\n",
       "      <td>المعی</td>\n",
       "      <td>/publisher/1589-%d8%a7%d9%84%d9%85%d8%b9%db%8c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1773</td>\n",
       "      <td>کارگاه فیلم و گرافیک سپاس</td>\n",
       "      <td>/publisher/1773-%da%a9%d8%a7%d8%b1%da%af%d8%a7...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>85</td>\n",
       "      <td>نیلا</td>\n",
       "      <td>/publisher/85-%d9%86%db%8c%d9%84%d8%a7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>1142</td>\n",
       "      <td>ناهید</td>\n",
       "      <td>/publisher/1142-%d9%86%d8%a7%d9%87%db%8c%d8%af</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>1346</td>\n",
       "      <td>بدرقه جاویدان</td>\n",
       "      <td>/publisher/1346-%d8%a8%d8%af%d8%b1%d9%82%d9%87...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>63</td>\n",
       "      <td>نفیر</td>\n",
       "      <td>/publisher/63-%d9%86%d9%81%db%8c%d8%b1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>1183</td>\n",
       "      <td>دبیر</td>\n",
       "      <td>/publisher/1183-%d8%af%d8%a8%db%8c%d8%b1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>150 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       id                       name  \\\n",
       "0      72                   کتاب آمه   \n",
       "1    1588                        اشک   \n",
       "2    1217               دنیای اقتصاد   \n",
       "3    1589                      المعی   \n",
       "4    1773  کارگاه فیلم و گرافیک سپاس   \n",
       "..    ...                        ...   \n",
       "211    85                       نیلا   \n",
       "212  1142                      ناهید   \n",
       "216  1346              بدرقه جاویدان   \n",
       "217    63                       نفیر   \n",
       "219  1183                       دبیر   \n",
       "\n",
       "                                                  link  \n",
       "0    /publisher/72-%da%a9%d8%aa%d8%a7%d8%a8-%d8%a2%...  \n",
       "1                   /publisher/1588-%d8%a7%d8%b4%da%a9  \n",
       "2    /publisher/1217-%d8%af%d9%86%db%8c%d8%a7%db%8c...  \n",
       "3       /publisher/1589-%d8%a7%d9%84%d9%85%d8%b9%db%8c  \n",
       "4    /publisher/1773-%da%a9%d8%a7%d8%b1%da%af%d8%a7...  \n",
       "..                                                 ...  \n",
       "211             /publisher/85-%d9%86%db%8c%d9%84%d8%a7  \n",
       "212     /publisher/1142-%d9%86%d8%a7%d9%87%db%8c%d8%af  \n",
       "216  /publisher/1346-%d8%a8%d8%af%d8%b1%d9%82%d9%87...  \n",
       "217             /publisher/63-%d9%86%d9%81%db%8c%d8%b1  \n",
       "219           /publisher/1183-%d8%af%d8%a8%db%8c%d8%b1  \n",
       "\n",
       "[150 rows x 3 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table_of_publishers = pd.DataFrame(publishers_data_list).drop_duplicates(subset=['id', 'name', 'link'])\n",
    "table_of_publishers.to_csv('./publishers.csv', index=False)\n",
    "table_of_publishers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c0d72ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>link</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>43502</td>\n",
       "      <td>آدام گرینفیلد</td>\n",
       "      <td>/profile/43502-adam-greenfield</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>23939</td>\n",
       "      <td>عباس مهرپویا</td>\n",
       "      <td>/profile/23939-abbas-mehrpouya</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4936</td>\n",
       "      <td>کنت بلانچارد</td>\n",
       "      <td>/profile/4936-kenneth-h-blanchard</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>39026</td>\n",
       "      <td>حسن رضوانی</td>\n",
       "      <td>/profile/39026-%d8%ad%d8%b3%d9%86-%d8%b1%d8%b6...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>68029</td>\n",
       "      <td>آلکسی آربوزوف</td>\n",
       "      <td>/profile/68029-aleksei-arbuzov</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>9801</td>\n",
       "      <td>بی جی نواک</td>\n",
       "      <td>/profile/9801-b-j-novak</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>4682</td>\n",
       "      <td>بلقیس سلیمانی</td>\n",
       "      <td>/profile/4682-belgheys-soleimani</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>22975</td>\n",
       "      <td>محمد رحمانیان</td>\n",
       "      <td>/profile/22975-%d9%85%d8%ad%d9%85%d8%af-%d8%b1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>9113</td>\n",
       "      <td>جیمز تربر</td>\n",
       "      <td>/profile/9113-james-thurber</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>364</td>\n",
       "      <td>مارگارت میچل</td>\n",
       "      <td>/profile/364-margaret-mitchell</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>187 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        id           name                                               link\n",
       "0    43502  آدام گرینفیلد                     /profile/43502-adam-greenfield\n",
       "1    23939   عباس مهرپویا                     /profile/23939-abbas-mehrpouya\n",
       "2     4936   کنت بلانچارد                  /profile/4936-kenneth-h-blanchard\n",
       "3    39026     حسن رضوانی  /profile/39026-%d8%ad%d8%b3%d9%86-%d8%b1%d8%b6...\n",
       "4    68029  آلکسی آربوزوف                     /profile/68029-aleksei-arbuzov\n",
       "..     ...            ...                                                ...\n",
       "202   9801     بی جی نواک                            /profile/9801-b-j-novak\n",
       "203   4682  بلقیس سلیمانی                   /profile/4682-belgheys-soleimani\n",
       "205  22975  محمد رحمانیان  /profile/22975-%d9%85%d8%ad%d9%85%d8%af-%d8%b1...\n",
       "206   9113      جیمز تربر                        /profile/9113-james-thurber\n",
       "209    364   مارگارت میچل                     /profile/364-margaret-mitchell\n",
       "\n",
       "[187 rows x 3 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table_of_writers = pd.DataFrame(writers_data_list).drop_duplicates(subset=['id', 'name', 'link'])\n",
    "table_of_writers.to_csv('./writers.csv', index=False)\n",
    "table_of_writers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a8c3505",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
